src/sentry/sentry_apps/token_exchange/grant_exchanger.py

```diff
@@ -31,15 +34,25 @@ class GrantExchanger:
 
     def run(self):
         with transaction.atomic(using=router.db_for_write(ApiToken)):
-            self._validate()
-            token = self._create_token()
-
-            # Once it's exchanged it's no longer valid and should not be
-            # exchangeable, so we delete it.
-            self._delete_grant()
-            self.record_analytics()
+            try:
+                self._validate()
+                token = self._create_token()
+
+                # Once it's exchanged it's no longer valid and should not be
+                # exchangeable, so we delete it.
+                self._delete_grant()
+            except APIUnauthorized:
+                logger.info(
+                    "grant-exchanger.context",
+                    extra={
+                        "application_id": self.application.id,
+                        "grant_id": self.grant.id,
+                    },
+                )
+                raise
```

Comment:
Main goal is trying to get some attributes that only exist in the refresher/grant_exchange scope. One approach ^
Alternatively could add context to exceptions with [add_note](https://docs.python.org/3/tutorial/errors.html#enriching-exceptions-with-notes) 
e.g
```       except ApiToken.DoesNotExist as e:
                    e.add_note(f"application_id:  {self.application.id}")
                    e.add_note(f"grant_id:  {self.self.grant.id}")
                    raise APIUnauthorized("Token does not exist")  from e
```